crates/ruff_python_codegen/src/generator.rs

```diff
@@ -1190,6 +1191,7 @@ impl<'a> Generator<'a> {
             Expr::IpyEscapeCommand(ast::ExprIpyEscapeCommand { kind, value, .. }) => {
                 self.p(&format!("{kind}{value}"));
             }
+            Expr::Invalid(ast::ExprInvalid { value, .. }) => self.p(value.as_str()),
```

Comment:
For this case, the resulting AST will be
```
Module(
        ModModule {
            range: 0..43,
            body: [
                Assign(
                    StmtAssign {
                        range: 1..20,
                        targets: [
                            Name(
                                ExprName {
                                    range: 1..2,
                                    id: "a",
                                    ctx: Store,
                                },
                            ),
                        ],
                        value: BinOp(
                            ExprBinOp {
                                range: 6..15,
                                left: NumberLiteral(
                                    ExprNumberLiteral {
                                        range: 6..7,
                                        value: Int(
                                            4,
                                        ),
                                    },
                                ),
                                op: Add,
                                right: NumberLiteral(
                                    ExprNumberLiteral {
                                        range: 14..15,
                                        value: Int(
                                            4,
                                        ),
                                    },
                                ),
                            },
                        ),
                    },
                ),
            ],
        },
    ),
```
Notice that only the binary operation appears, this is due to when parsing a parenthesized expression (or an expression inside `[]`, `{}`)  with unexpected tokens, in this case it would be `ident` (`4ident` is lexed as two different tokens), `===` and `bcd`, the parser will skip these tokens in an attempt to find the `)`. So the reprint it's probably even worse 🤣, because it'll generate this  `a = 4 + 4`. 

The error message needs improvement for this case, because it doesn't create `unexpected token` error messages, it only creates the `expecting ")", found "ident"` error message.

Reply:
Hmm that's a serious problem. It's okay for the parser to skip tokens but these tokens must be attached to the AST. I assumed it would do so by creating an `InvalidExpr` for the invalid nodes. Let me spends some time today to take a closer look at the parser's error recovery.

Reply:
Probably won't be much work to create an `InvalidExpr` for the skipped tokens.